express_server.js

```diff
@@ -6,135 +6,175 @@ const cookieSession = require("cookie-session");
+    res.status(406).redirect("/errors/406");
   } else if (getUserByEmail(req.body.email, users)) {
-    res.send("Hey! We're already friends, just <a href='/login'>log in!</a> :)");
+    res.status(409).redirect("/errors/409");
   } else {
-    users[newUserID] = {id: newUserID, email: req.body.email, password: bcrypt.hashSync(req.body.password, 10)};
+    const newUserID = generateID();
+    users[newUserID] = {
+      id: newUserID,
+      email: req.body.email,
+      password: hashPassword(req.body.password),
+    };
     req.session.user_id = newUserID;
     res.redirect("/urls");
   }
 });
 
 //renders register template
-app.get('/register', function(req, res) {
-  let templateVars = { urls: urlDatabase, user:users[req.session.user_id] };
+app.get("/register", (req, res) => {
+  const templateVars = { user: req.session.user_id };
   res.render("user_registration", templateVars);
 });
 
-app.post('/login', function(req, res) {
-  if (!req.body.email || !req.body.password) {
-    // errors if fields are empty
-    res.status(400).send("Oops! Please enter an email and password :) <a href='/login'>Login</a>");
-  } else if (getUserByEmail(req.body.email, users) === undefined) {
-    // errors if user is not registered
-    res.status(403).send("Looks like we're not friends yet :( This email does not exist, please <a href='/register'>register!</a>");
-  } else if (getUserByEmail(req.body.email, users) && !bcrypt.compareSync(req.body.password, users[getUserByEmail(req.body.email, users)].password)) {
-    // errors if user is registered and passwords do not match
-    res.status(403).send("Oops! Password and email do not match! <a href='/login'>Login</a>");
-  } else if (getUserByEmail(req.body.email, users) && bcrypt.compareSync(req.body.password, users[getUserByEmail(req.body.email, users)].password)) {
+app.post("/login", (req, res) => {
+  const email = req.body.email;
+  const password = req.body.password;
+  if (
+    emailExists(users, email) &&
+    areValidEmailPassword(users, email, password)
+  ) {
+    // set the cookie
     req.session.user_id = getUserByEmail(req.body.email, users);
     res.redirect("/urls");
+  } else {
+    res.status(400).redirect("/errors/400");
   }
 });
 
 //renders the user log-in page
-app.get('/login/', function(req, res) {
-  let templateVars = { urls: urlDatabase, user:users[req.session.user_id] };
+app.get("/login/", (req, res) => {
+  let templateVars = { user: req.session.user_id };
   res.render("user_login", templateVars);
 });
 
-app.get("/logout", function(req, res) {
+app.get("/logout", (req, res) => {
   req.session = null;
-  res.redirect('/urls');
+  res.redirect("/login");
 });
 
 // main page
 app.get("/", (req, res) => {
-  let templateVars = { urls: userURLs(req.session.user_id, urlDatabase), user: users[req.session.user_id] };
-  if (templateVars.user === undefined) {
-    res.redirect("/login");
-  } else {
-    res.redirect("/urls");
-  }
+  res.redirect("/urls");
```

Comment:
This endpoint was doing exactly the same as `/urls` so you can just redirect to it and get rid of the duplication.